### PR TITLE
Add StopHandler object to runtime langlib for graceful shutdown of modules

### LINKED ISSUE #36207: Add lang lib to register stop handler which should be run with graceful stop
**Description:**
Implementation of https://github.com/ballerina-platform/ballerina-spec/issues/730

### LINKED ISSUE #36401: Graceful shutdown of listeners fail with CCE when registering more than one listeners with a module or listeners registered with imported modules
**Description:**

$title

**Steps to reproduce:**
case 1: two listeners registered with a module and `gracefulStop()` method has an async call inside it.  
```ballerina
import ballerina/io;
import ballerina/lang.runtime;


public class ListenerObj1 {
    public function init() {    
    }

    public function 'start() {    
    }

    public function gracefulStop() {
        runtime:sleep(5);
        io:println("graceful stop of Listener1");
    }

    public function immediateStop() {    
    }
}

final ListenerObj1 lo1 = new ListenerObj1();


public class ListenerObj2 {
    public function init() {}

    public function 'start() {}

    public function gracefulStop() {
        runtime:sleep(5);
        io:println("graceful stop of Listener2");
    }

    public function immediateStop() {}
}

final ListenerObj2 lo2 = new ListenerObj2();

public function main() {
    runtime:registerListener(lo1);
    runtime:registerListener(lo2);
    bar();
}

function foo() {
    io:println("foo...");
    lo2.gracefulStop();
    runtime:deregisterListener(lo2);
    io:println("foo after deregister lo2\n");
}

function bar() {
    io:println("bar...");
    lo1.gracefulStop();
    runtime:deregisterListener(lo1);
    io:println("bar after deregister lo1\n");
    foo();
}
```
case 2: Imported modules also have listeners registered with them
moduleA#main.bal
```ballerina
import ballerina/io;
import ballerina/lang.runtime;


public class ListenerA {
    public function init() {    
    }

    public function 'start() {    
    }

    public function gracefulStop() {
        runtime:sleep(5);
        io:println("graceful stop of ListenerA");
    }

    public function immediateStop() {    
    }
}

ListenerA listenerA = new ListenerA();

public function main() {
    runtime:registerListener(listenerA);
    func();
}

function func() {
    io:println("func...");
    listenerA.gracefulStop();
    runtime:deregisterListener(listenerA);
}
```
main.bal
```ballerina
import ballerina/io;
import mysql_client.moduleA;
import ballerina/lang.runtime;


public class ListenerObj1 {
    public function init() {    
    }

    public function 'start() {    
    }

    public function gracefulStop() {
        runtime:sleep(5);
        io:println("graceful stop of Listener1");
    }

    public function immediateStop() {    
    }
}

final ListenerObj1 lo1 = new ListenerObj1();

public function main() {
    runtime:registerListener(lo1);
    bar();
}

function bar() {
    lo1.gracefulStop();
    runtime:deregisterListener(lo1);
    io:println("bar after deregister lo1\n");
    moduleA:main();
}
```

**Affected Versions:**

**OS, DB, other environment details and versions:**

**Related Issues (optional):**

**Suggested Labels (optional):**

**Suggested Assignees (optional):**

### PR DESCRIPTION
**Pull request by gabilang**
Fixes #36207 
Fixes #36401 

## Check List 
- [x] Read the [Contributing Guide](https://github.com/ballerina-platform/ballerina-lang/blob/master/CONTRIBUTING.md)
- [ ] Updated Change Log
- [ ] Checked Tooling Support (#<Issue Number>)
- [ ] Added necessary tests
  - [x] Unit Tests
  - [ ] Spec Conformance Tests
  - [ ] Integration Tests
  - [ ] Ballerina By Example Tests
- [ ] Increased Test Coverage   
- [ ] Added necessary documentation  
  - [ ] API documentation 
  - [ ] Module documentation in Module.md files
  - [ ] Ballerina By Examples

### CODE REVIEW

**Comment by warunalakshitha** [comment]
Lets add test with multiple module register stop handlers, register same stop handler multiple times and register stop handler inside functions.

**Comment by warunalakshitha** [comment]
Also a test with panic cases as well.

File: tests/jballerina-integration-test/src/test/resources/packaging/dynamic_listener_async_call_test_project/main.bal [hunk 1]
```diff
@@ -0,0 +1,73 @@
+// Copyright (c) 2022 WSO2 Inc. (http://www.wso2.org) All Rights Reserved.
```
Comment (warunalakshitha) [inline tests/jballerina-integration-test/src/test/resources/packaging/dynamic_listener_async_call_test_project/main.bal:1]: Lets add test with multiple module register stop handlers, register same stop handler multiple times and register stop handler inside functions.

File: tests/jballerina-integration-test/src/test/resources/packaging/dynamic_listener_async_call_test_project/main.bal [hunk 2]
```diff
@@ -0,0 +1,73 @@
...
+//
+// WSO2 Inc. licenses this file to you under the Apache License,
```
Comment (warunalakshitha) [inline tests/jballerina-integration-test/src/test/resources/packaging/dynamic_listener_async_call_test_project/main.bal:3]: Also a test with panic cases as well.

File: bvm/ballerina-runtime/src/main/java/io/ballerina/runtime/internal/scheduling/RuntimeRegistry.java [hunk 3]
```diff
@@ -0,0 +1,151 @@
...
+    }
+
+    public synchronized void deregisterListener(BObject listener) {
+        listenerSet.remove(listener);
+        if (!scheduler.isListenerDeclarationFound() && listenerSet.isEmpty()) {
+            scheduler.setImmortal(false);
+        }
+    }
+
+    public synchronized void stopListeners(Strand strand) {
```
Comment (warunalakshitha) [inline bvm/ballerina-runtime/src/main/java/io/ballerina/runtime/internal/scheduling/RuntimeRegistry.java:63]: Lets add stopdynamic and merge both listener graceful stop and stop handler graceful stop

**Comment by gabilang** [comment]
Closing this since the commits are added to the PR https://github.com/ballerina-platform/ballerina-lang/pull/36574

**Pull request closed without merging**